src/Database/Bloodhound/Types.hs

```diff
@@ -604,7 +605,14 @@ data Search = Search { queryBody       :: Maybe Query
                        -- default False
                      , trackSortScores :: TrackSortScores
                      , from            :: From
-                     , size            :: Size } deriving (Eq, Show)
+                     , size            :: Size
+                     , source          :: Maybe Source } deriving (Eq, Show)
+
+data Source =
+  NoSource
+  | SourceFields FieldOrFields
+  | SourceIncludeExclude [Text] [Text]
```

Comment:
I'd go a step further:

```haskell
newtype Pattern = Pattern Text

data Include = Include [Pattern]
data Exclude = Exclude [Pattern]
```

This leaves the door open for having a better typed grammar for patterns later (an AST'ish thing) and makes the specification more self-documenting.